File: libmuscle/python/libmuscle/port_manager.py
from typing import Dict, List, Optional, Tuple

from ymmsl import Identifier, Operator

from libmuscle.peer_info import PeerInfo
from libmuscle.port import Port


class PortManager:
    """Manages sending and receiving ports of the current instance."""
    def __init__(
            self, index: List[int], declared_ports: Optional[Dict[Operator, List[str]]]
            ) -> None:
        """Create a PortManager.

        Args:
            index: The index of this instance.
            declared_ports: The ports this instance has, as declared by the
                    user when creating the Instance object.
        """
        self._index = index
        self._ports: Dict[str, Port] = {}
        self._declared_ports = declared_ports

    def connect_ports(self, peer_info: PeerInfo) -> None:
        """Connect the ports to their peers.

        This is the second stage in the simulation wiring process.

        Peers here are instances, and the information about them received
        from the manager is in peer_info. We are going to create a set of
        Port objects here, one for each of our ports.

        If the user gave us a set of ports (i.e. they were declared in the
        code) then we'll create a Port object for each of those, with
        information about the attached conduit / peer (if any) from
        peer_info. If the user did not give us any ports (which is legal),
        then the ports will be created entirely from the information
        received from the manager.  The user then has to use
        Instance.list_ports() to see what ports they got, and do something
        with them.

        Args:
            peer_info: Information about our peers from the manager.
        """
        self._muscle_settings_in = self._settings_in_port(peer_info)
        if self._declared_ports:
            self._ports = self._ports_from_declared(peer_info)
        else:
            self._ports = self._ports_from_conduits(peer_info)

    def settings_in_connected(self) -> bool:
        """Returns whether muscle_settings_in is connected."""
        return self._muscle_settings_in.is_connected()

    def list_ports(self) -> Dict[Operator, List[str]]:
        """Returns a description of the ports this PortManager has.

        Returns:
            A dictionary, indexed by Operator, containing lists of
            port names. Operators with no associated ports are not
            included.
        """
        result: Dict[Operator, List[str]] = {}
        for port_name, port in self._ports.items():
            if port.operator not in result:
                result[port.operator] = list()
            result[port.operator].append(port_name)
        return result

    def port_exists(self, port_name: str) -> bool:
        """Returns whether a port with the given name exists.

        Args:
            port_name: Port name to check.

        Returns:
            True iff the port exists
        """
        return port_name in self._ports

    def get_port(self, port_name: str) -> Port:
        """Returns a Port object describing a port with the given name.

        Args:
            port_name: Name of the port to retrieve.

        Returns:
            A Port object for the port
        """
        return self._ports[port_name]

    def restore_message_counts(
            self, port_message_counts: Dict[str, List[int]]) -> None:
        """Restore message counts on all ports.

        Args:
            port_message_counts: The message counts, as a dictionary indexed by
                    port name containing a list of counts, one for each slot.
        """
        for port_name, num_messages in port_message_counts.items():
            if port_name == "muscle_settings_in":
                self._muscle_settings_in.restore_message_counts(num_messages)
            elif port_name in self._ports:
                self._ports[port_name].restore_message_counts(num_messages)
            else:
                raise RuntimeError(f'Unknown port {port_name} in snapshot.'
                                   ' Have your port definitions changed since'
                                   ' the snapshot was taken?')

    def get_message_counts(self) -> Dict[str, List[int]]:
        """Get message counts for all ports on the communicator.

        Returns:
            A dictionary indexed by port name containing a list of counts, one
            for each slot of the corresponding port.
        """
        port_message_counts = {port_name: port.get_message_counts()
                               for port_name, port in self._ports.items()}
        port_message_counts["muscle_settings_in"] = \
            self._muscle_settings_in.get_message_counts()
        return port_message_counts

    def _settings_in_port(self, peer_info: PeerInfo) -> Port:
        """Creates a Port representing muscle_settings_in.

        Args:
            peer_info: Information about our peers from the manager.
        """
        msi = Identifier('muscle_settings_in')
        if peer_info.is_connected(msi):
            sender_component = peer_info.get_peer_ports(msi)[0][:-1]
            return Port(
                    str(msi), Operator.F_INIT, False, True, len(self._index),
                    peer_info.get_peer_dims(sender_component))

        return Port(
                str(msi), Operator.F_INIT, False, False, len(self._index), [])

    def _ports_from_declared(self, peer_info: PeerInfo) -> Dict[str, Port]:
        """Derives port definitions from supplied declaration.

        Args:
            peer_info: Information about our peers from the manager.

        Returns:
            A dictionary keyed by port name containing corresponding Port
                    objects.
        """
        assert self._declared_ports is not None

        ports = dict()
        for operator, port_list in self._declared_ports.items():
            for port_desc in port_list:
                port_name, is_vector = self._split_port_desc(port_desc)
                if port_name.startswith('muscle_'):
                    raise RuntimeError(
                            'Port names starting with "muscle_" are reserved for'
                            ' MUSCLE, please rename port "{}"'.format(port_name))
                port_id = Identifier(port_name)
                is_connected = peer_info.is_connected(port_id)
                if is_connected:
                    peer_ports = peer_info.get_peer_ports(port_id)
                    peer_port = peer_ports[0]
                    peer_component = peer_port[:-1]
                    port_peer_dims = peer_info.get_peer_dims(peer_component)
                    for peer_port in peer_ports[1:]:
                        peer_component = peer_port[:-1]
                        if port_peer_dims != peer_info.get_peer_dims(peer_component):
                            port_strs = ', '.join(map(str, peer_ports))
                            raise RuntimeError(
                                    'Multicast port "{}" is connected to peers with'
                                    ' different dimensions. All peer components that'
                                    ' this port is connected to must have the same'
                                    ' multiplicity. Connected to ports: {}.'.format(
                                        port_name, port_strs))
                else:
                    port_peer_dims = []
                ports[port_name] = Port(
                        port_name, operator, is_vector, is_connected,
                        len(self._index), port_peer_dims)
        return ports

    def _ports_from_conduits(self, peer_info: PeerInfo) -> Dict[str, Port]:
        """Derives port definitions from conduits.

        Args:
            peer_info: Information about our peers from the manager.
        """
        ports = dict()

        def make_port(
                port_id: Identifier, operator: Operator, peer_dims: List[int]) -> None:
            if not str(port_id).startswith('muscle_'):
                ndims = max(0, len(peer_dims) - len(self._index))
                is_vector = (ndims == 1)
                is_connected = peer_info.is_connected(port_id)
                ports[str(port_id)] = Port(
                        str(port_id), operator, is_vector, is_connected,
                        len(self._index), peer_dims)

        for port_id, sender_ref in peer_info.list_incoming_ports():
            peer_dims = peer_info.get_peer_dims(sender_ref[:-1])
            make_port(port_id, Operator.F_INIT, peer_dims)

        for port_id, receiver_refs in peer_info.list_outgoing_ports():
            peer_dims = peer_info.get_peer_dims(receiver_refs[0][:-1])
            make_port(port_id, Operator.O_F, peer_dims)

        return ports

    def _split_port_desc(self, port_desc: str) -> Tuple[str, bool]:
        """Split a port description into its name and dimensionality.

        Expects a port description of the form port_name or
        port_name[], and returns the port name and whether it is a
        vector port.

        Args:
            port_desc: A port description string, as above.
        """
        is_vector = False
        if port_desc.endswith('[]'):
            is_vector = True
            port_desc = port_desc[:-2]

        if port_desc.endswith('[]'):
            raise ValueError(('Port description "{}" is invalid: ports can'
                              ' have at most one dimension.').format(
                                  port_desc))

        return port_desc, is_vector

File: libmuscle/cpp/src/libmuscle/port_manager.cpp
#include <libmuscle/port_manager.hpp>

#include <cassert>
#include <limits>
#include <sstream>
#include <tuple>


using ymmsl::Conduit;
using ymmsl::Identifier;
using ymmsl::Operator;
using ymmsl::Reference;

using std::get;
using std::prev;


namespace libmuscle { namespace _MUSCLE_IMPL_NS {

PortManager::PortManager(
        std::vector<int> const & index,
        Optional<PortsDescription> const & declared_ports)
    : index_(index)
    , declared_ports_(declared_ports)
    , ports_()
{}

void PortManager::connect_ports(PeerInfo const & peer_info) {
    muscle_settings_in_ = settings_in_port_(peer_info);

    if (declared_ports_.is_set())
        ports_ = ports_from_declared_(peer_info);
    else
        ports_ = ports_from_conduits_(peer_info);
}

bool PortManager::settings_in_connected() const {
    return muscle_settings_in_.get().is_connected();
}

Port const & PortManager::muscle_settings_in() const {
    return muscle_settings_in_.get();
}

Port & PortManager::muscle_settings_in() {
    return muscle_settings_in_.get();
}

PortsDescription PortManager::list_ports() const {
    PortsDescription result;
    for (auto const & port : ports_)
        result[port.second.oper].emplace_back(port.first);
    return result;
}

bool PortManager::port_exists(std::string const & port_name) const {
    return ports_.count(port_name) > 0;
}

Port const & PortManager::get_port(std::string const & port_name) const {
    return ports_.at(port_name);
}

Port & PortManager::get_port(std::string const & port_name) {
    return ports_.at(port_name);
}

PortManager::PortMessageCounts PortManager::get_message_counts() {
    PortMessageCounts port_message_counts;
    for(auto const & port_item : ports_)
        port_message_counts[port_item.first] = port_item.second.get_message_counts();

    assert(muscle_settings_in_.is_set());  // is always created by connect_ports()
    auto counts = muscle_settings_in_.get().get_message_counts();
    port_message_counts["muscle_settings_in"] = counts;

    return port_message_counts;
}

void PortManager::restore_message_counts(
        PortManager::PortMessageCounts const & port_message_counts) {
    for (auto const & item : port_message_counts) {
        if (item.first == "muscle_settings_in") {
            assert(muscle_settings_in_.is_set());  // is always created by connect()
            muscle_settings_in_.get().restore_message_counts(item.second);
        } else {
            auto port_item = ports_.find(item.first);
            if (port_item != ports_.end()) {
                port_item->second.restore_message_counts(item.second);
            } else {
                throw std::runtime_error(
                        "Unknown port " + item.first + " in snapshot."
                        " Have your port definitions changed since"
                        " the snapshot was taken?");
            }
        }
    }
}


PortManager::Ports_ PortManager::ports_from_declared_(PeerInfo const & peer_info) const {
    Ports_ ports;
    for (auto const & ppo : declared_ports_.get()) {
        for (auto const & port_desc : ppo.second) {
            std::string port_name;
            bool is_vector;
            std::tie(port_name, is_vector) = split_port_desc_(port_desc);
            if (port_name.find("muscle_") == 0u) {
                std::ostringstream oss;
                oss << "Port names starting with 'muscle_' are reserved for";
                oss << " MUSCLE, please rename port '" << port_name << "'.";
                throw std::runtime_error(oss.str());
            }
            bool is_connected = peer_info.is_connected(port_name);
            std::vector<int> port_peer_dims;
            if (is_connected) {
                auto peer_ports = peer_info.get_peer_ports(port_name);
                Reference peer_port = peer_ports.at(0);
                Reference peer_component(
                        peer_port.cbegin(), std::prev(peer_port.cend()));
                port_peer_dims = peer_info.get_peer_dims(peer_component);
                for (std::size_t i = 1; i < peer_ports.size(); i++) {
                    peer_port = peer_ports.at(i);
                    peer_component = Reference(
                            peer_port.cbegin(), std::prev(peer_port.cend()));
                    if (port_peer_dims != peer_info.get_peer_dims(peer_component)) {
                        std::stringstream ss;
                        ss << "Multicast port \"" << port_name;
                        ss << "\" is connected to peers with different";
                        ss << " dimensions. All peer components that this";
                        ss << " port is connected to must have the same";
                        ss << " multiplicity. Connected to ports: ";
                        bool first = true;
                        for (auto port : peer_ports) {
                            if (first)
                                first = false;
                            else
                                ss << ", ";
                            ss << port;
                        }
                        throw std::runtime_error(ss.str());
                    }
                }
            }
            ports.emplace(port_name, Port(
                    port_name, ppo.first, is_vector, is_connected,
                    index_.size(), port_peer_dims));
        }
    }
    return ports;
}

PortManager::Ports_ PortManager::ports_from_conduits_(PeerInfo const & peer_info) const {
    Ports_ ports;

    auto make_port = [&](
            Identifier const & port_id, Operator oper,
            std::vector<int> const & peer_dims) {
        int ndims = std::max(std::vector<int>::size_type(0u), peer_dims.size() - index_.size());
        bool is_vector = (ndims == 1);
        bool is_connected = peer_info.is_connected(port_id);
        if (std::string(port_id).find("muscle_") != 0u) {
            ports.emplace(port_id, Port(
                    port_id, oper, is_vector, is_connected, index_.size(), peer_dims));
        }
    };

    for (auto const & port_sender : peer_info.list_incoming_ports()) {
        std::string port_name(get<0>(port_sender));
        Reference sending_component(
                get<1>(port_sender).cbegin(), prev(get<1>(port_sender).cend()));

        auto const & peer_dims = peer_info.get_peer_dims(sending_component);
        make_port(get<0>(port_sender), Operator::F_INIT, peer_dims);
    }

    for (auto const & port_recvs : peer_info.list_outgoing_ports()) {
        std::string port_name(get<0>(port_recvs));
        Reference receiving_component(
                get<1>(port_recvs)[0].cbegin(),
                prev(get<1>(port_recvs)[0].cend()));

        auto const & peer_dims = peer_info.get_peer_dims(receiving_component);
        make_port(get<0>(port_recvs), Operator::O_F, peer_dims);
    }

    return ports;
}

Port PortManager::settings_in_port_(PeerInfo const & peer_info) const {
    Identifier msi("muscle_settings_in");
    if (peer_info.is_connected(msi)) {
            Reference sender_port = peer_info.get_peer_ports(msi)[0];
            Reference sender_component(
                    sender_port.cbegin(), std::prev(sender_port.cend()));
            return Port(
                    std::string(msi), Operator::F_INIT, false, true, index_.size(), {});
    }
    return Port(std::string(msi), Operator::F_INIT, false, false, index_.size(), {});
}

std::tuple<std::string, bool> PortManager::split_port_desc_(
        std::string const & port_desc) const {
    std::string port_name(port_desc);
    bool is_vector = false;

    auto found = port_desc.rfind("[]");
    if (found != std::string::npos && found == (port_desc.size() - 2)) {
        is_vector = true;
        port_name = port_desc.substr(0, port_desc.size() - 2);
    }

    found = port_name.rfind("[]");
    if (found != std::string::npos && found == (port_name.size() - 2)) {
        std::ostringstream oss;
        oss << "Port description '" << port_desc << "' is invalid: ports can";
        oss << " have at most one dimension.";
        throw std::invalid_argument(oss.str());
    }

    return std::make_tuple(port_name, is_vector);
}

} }

